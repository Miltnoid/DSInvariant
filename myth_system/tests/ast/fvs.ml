type id =
  | A
  | B
  | C

type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of id * list

type exp =
  | EBVar of nat
  | EFVar of id
  | EApp  of exp * exp

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (hd, tl) -> Cons (hd, append tl l2)
;;

let fvs : exp -> list |>
  { EBVar (0) => []
  | EFVar (A) => [A]
  | EFVar (B) => [B]
  | EApp (EBVar (0), EBVar (0)) => []
  | EApp (EFVar (A), EBVar (0)) => [A]
  | EApp (EBVar (0), EFVar (A)) => [A]
  } = ?
