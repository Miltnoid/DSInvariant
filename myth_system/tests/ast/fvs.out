(* Generated Wed Dec 31 11:17:18 2014 *)
let fvs : exp -> list =
  let rec f1 (e1:exp) : list =
    match e1 with
      | EBVar n1 -> Nil
      | EFVar i1 -> Cons (i1, Nil)
      | EApp (e2, e3) -> append (f1 e3) (f1 e2)
  in
    f1
;;
